KIPs/kip-114.md

```diff
@@ -0,0 +1,324 @@
+    if header.number == FORK_BLOCK:
+        # parent mixHash does not exist, so fill with default
+        parentMixHash = b"\x00" * 32
+    mixHash = calc_mix_hash(parentMixHash, header.randomReveal)
+    return mixHash == header.mixHash
+
+
+def main():
+    header = Header(FORK_BLOCK - 2, b"\x04" * 96, b"\x00" * 32)
+    prevHeader = header
+    N = 10
+    # print header numbers in [FORK_BLOCK - 1, FORK_BLOCK + N]
+    for _ in range(N + 2):
+        header = gen_next_header(header)
+        print(header)
+        verified = verify_xxxx_header(header, prevHeader.mixHash)
+        assert verified
+        print("verified:", verified)
+        print("=" * 80)
+        prevHeader = header
+
+
+main()
+```
+
+### Genesis Block
+
+If `FORK_BLOCK` is zero, the genesis block must have `randomReveal`, `mixHash` fields.
+
+## Rationale
+
+### Names of the header fields
+
+The naming is highly inspired by [EIP-4339](https://eips.ethereum.org/EIPS/eip-4399).
+We use `randomReveal` instead of RANDAO reveal because the randoms are generated in a different manner.
+
+### Block size increment due to new fields in the header
+
+The size increment cost of RLP-encoded header is around 131 bytes per block, which is approximately 3.8GB per year. It is a negligible cost compared to its benefit.
+
+### BLS-based random over EC-based random
+
+We chose BLS-based verifiable random over EC-based because:
+
+- A BLS library written in Golang is easy to find as in [link](https://github.com/klaytn/klaytn/issues/1823), whereas it is not the case for EC-based as in [link](https://www.ietf.org/archive/id/draft-irtf-cfrg-vrf-15.html#section-6).
+- There is a potential consensus change based on BLS signature. Therefore, this proposal gently introduces BLS signature to the Klaytn protocol as well as adding on-chain randomness.
+
+## Backwards Compatibility
+
+Blocks before `FORK_BLOCK` will remain the same.
+
+## Security Considerations
+
+### Biasability
+
+BLS signature as well as the signing message (i.e., proposing block number) is deterministic. Thus, the proposer has no influence over the computation of `mixHash`. In addition, the biasability of `mixHash` mainly depends on that of the hash of `randomReveal`. To the best of our knowledge, keccak256 is not a biased hash function. Therefore, `mixHash`, which is the XOR of the output of keccak256 and the previous `mixHash`, is unbiased.
+
+### Predictability
+
+`mixHash` of the next block can be computed when the both requirements are met: (1) the `mixHash` of the latest block is revealed; and (2) the `randomReveal` of the next block is revealed. Even if the proposer of the next block can precompute the latter as soon as the proposer found out that it is proposing the next block, it does not know the former which circumvents the proposer from predicting `mixHash` of the next block. In other words, only after the proposer receives the latest `mixHash`, it is able to know the `mixHash` of the next block.
```

Comment:
hmm the sentences are little bit hard to understand.
1. "next" is shown several times, but the current block is unclear.
2. Anyway, the proposer of the block 101 will always know `mixHash` of block 100 before starting the mining. So, when the proposer is processing the block, it always know the `randomReveal` and `mixHash` of the block.

What we emphasize here is
1) `mixHash` cannot be tampered by the proposer of the previous block
2) `randomReveal` cannot be tampered by the proposer of the current block.

Since `randomReveal` is calculated from `prevBlock.mixHash` and the current proposer's signature of the current block number, so no one can predict the value. Even the current block proposer cannot predict the value.

Reply:
It was too redundant, so I shortened it. PTAL